### PR TITLE
feat: delete organization

### LINKED ISSUE #5737: Delete Organizations
At the moment the delete functionality in ZITADEL is implemented more as an "archive" function. As an administrator I want that when I click "delete" in the console or send "delete" to the API, that the organization is actually deleted.

```[tasklist]
### Acceptance Criteria
- [ ] I am able to delete an organization through the api or console
- [ ] A deleted organization is not returned in the list organizations request
- [ ] No resources of a deleted organization can be read or manipulated
- [ ] I can't manipulate a deleted organization
```

### PR DESCRIPTION
**Pull request by doncicuto**
At the moment the delete functionality in ZITADEL is implemented more as an "archive" function. As an administrator I want that when I click "delete" in the console or send "delete" to the API, that the organization is actually deleted.

Right now if we delete an org it's kept on the projections.org table with state 3 -> OrgStateRemoved. In the following screenshot we can see org Test, that has been deleted, has org_state set to 3. Also we see an org named Deleted that is active.

org_deleted_db_1

In this PR the org is removed from that table when the org is deleted as requested. We can see in the following screenshot that org Delete is no more there.

org_deleted_db_2

Also if a resource depends on that org (for example a project and a project role) we can check in the following screenshots that those resources are also deleted

Before delete, project Test and role Example are there associated with the org.

org_deleted_resources

After delete, project Test and role Example are no longer there.

org_deleted_resources_2

The following video shows also how to delete the org using the API and how the org is no longer retrieved from the list of available orgs:

https://github.com/zitadel/zitadel/assets/30386061/4819b786-b98b-40ae-8305-6647e09153d8

Some things that should be noted:

1. The OrgStateRemoved state hasn't been removed as we may have previous removed org in our Zitadel db instance and for a while we'd like to see the "Removed" tag in the console. Maybe a migration task could be added to remove those orgs in that state from the database and then later on in a different PR this old state could be removed. I don't include the migration task in this PR but if you think this is the right approach please create an issue and leave there some hints so I can create that migration task.
2. The owner_removed column is defined in several tables. This PR doesn't remove that column from thos tables as once again it may affect current Zitadel instances. As I don't know how the migration tasks should be prepared this should be noted in another issue or keep #5737 open with pending work. Once again if you give me some hints I could have a look at it :smile:

Acceptance Criteria

- [X] I am able to delete an organization through the api or console
- [X] A deleted organization is not returned in the list organizations request
- [X] No resources of a deleted organization can be read or manipulated
- [X] I can't manipulate a deleted organization

Should close #5737

### Definition of Ready

- [X] I am happy with the code
- [X] Short description of the feature/issue is added in the pr description
- [X] PR is linked to the corresponding user story
- [X] Acceptance criteria are met
- [ ] All open todos and follow ups are defined in a new ticket and justified
- [ ] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [X] No debug or dead code
- [X] My code has no repetitions
- [X] Critical parts are tested automatically
- [X] Where possible E2E tests are implemented
- [ ] Documentation/examples are up-to-date
- [ ] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.

### CODE REVIEW

**Review by muhlemmer — CHANGES_REQUESTED** [review]
Thanks for all the work! Just some comments. Appologies for taking so long to review, there are some merge conflicts by now. Can you please also resolve them?

> The owner_removed column is defined in several tables. This PR doesn't remove that column from thos tables as once again it may affect current Zitadel instances. As I don't know how the migration tasks should be prepared this should be noted in another issue or keep #5737 open with pending work. Once again if you give me some hints I could have a look at it

To migrate, we would delete the column definition from the projection in the go code and increase the version number of the table. However, please **do not** do this, because it will affect almost all projections and it will take lots of time and resources to Reduce all events in our production. We've discussed it internally and we will phase out the `owner_removed` column for each projection whenever another migration is needed.

File: internal/command/org_model.go
Comment (doncicuto) [reply internal/command/org_model.go:37]: Thanks a lot @muhlemmer for the explanation.

**Comment by doncicuto** [comment]
Thanks for your review @muhlemmer, I'll solve conflicts and apply your suggestions as soon as possible.

File: internal/query/org.go
Comment (doncicuto) [reply internal/query/org.go:212]: Hi @muhlemmer do you want me to change the version, the updated tests and remove the extra condition in this PR? 

I don't understand how changing the projection version will omit automatically any old row with org_state equals OrgStateRemoved, maybe it's a silly question but I'm still trying to grasp this eventstore/reduction magic :smile:

File: internal/command/org_model.go
Comment (doncicuto) [reply internal/command/org_model.go]: Fixed, thanks @muhlemmer

File: internal/query/org.go
Comment (muhlemmer) [reply internal/query/org.go:212]: Because the table is renamed, `setup` thinks it is missing. So a whole new table is created by the name  `projections.orgs1` and the events should be reduced into that one, using the updated reducers.

`projections.org` should still be there with reduced data corresponding the old reducing logic.

In a high availability like zitadel uses in production, this allows for a new version zitadel to be launched, create new projection(s) and populate them while an older version is still serving traffic using the old projection. Once the new projection is up-to-date production traffic can be redirected to the new version. If anything goes wrong, like a migration or unexpected errors in the new version, we can just make a switch back to the old version again.

So basically, we can operate multiple version of zitadel side-by-side from a database model perspective.

Comment (doncicuto) [reply internal/query/org.go:212]: Thanks again @muhlemmer for the explanation. I'll update the code removing the condition. I cant send the new code until next week due to lack of time.

**Comment by doncicuto** [comment]
Hi @hifabienne I'll try to finish it this weekend as well as other issues that I've left undone due to my holidayssss 😜

**Comment by doncicuto** [comment]
Hi @muhlemmer, I've changed projections.org version, updated the tests and removed the extra condition in SearchOrgs. Hope code is ready as needed. Requesting review again, thanks for your time.

**Review by muhlemmer — DISMISSED** [review]

**Comment by muhlemmer** [comment]
> Hi @muhlemmer, I've changed projections.org version, updated the tests and removed the extra condition in SearchOrgs. Hope code is ready as needed. Requesting review again, thanks for your time.

Thanks for YOUR time! Great work.

**Comment by muhlemmer** [comment]
It seems the e2e test fails on organisation deletions. I  imagine it still looks for the organistation with the "removed" status in the overview.

Do you know how to work the e2e tests @doncicuto? Myself I don't (I do only backend). But if you need some help, I will ping one of my colleagues for that.

**Comment by doncicuto** [comment]
Hi @muhlemmer, sorry I'll try to rerun e2e on my laptop this evening and let you know if I need some help, my fault. In the past GitHub run the tests as soon as I sent the PR but now I've to wait till you approve them.

**Comment by doncicuto** [comment]
@muhlemmer I managed to fix my org e2e tests. I'm having problems to run e2e tests in my laptop so my apologies, this time organization tests were succesful before I sent the PR. 

I've merged main into the PR so accidentaly I dismissed your review? (Github did) so I asked for a new review from you. 

Hope this time everything works fine and stop wasting your time. Cheers!

**Review by muhlemmer — APPROVED** [review]

**Pull request merged**
